File: 송하연/10월1주차/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {

        Map<String, Integer> genrePlay = new HashMap<>();
        Map<String, List<int[]>> genreInfo = new HashMap<>();
        
        //장르별 정렬
        int idx=0;
        for(String genre:genres){
            genrePlay.put(genre, genrePlay.getOrDefault(genre,0)+plays[idx]);
            
            genreInfo.putIfAbsent(genre, new ArrayList<>());
            genreInfo.get(genre).add(new int[]{idx,plays[idx]});
            
            idx++;
        }
        
        List<String> sortedGenre = new ArrayList<>(genrePlay.keySet());
        sortedGenre.sort((a,b)->genrePlay.get(b)-genrePlay.get(a));
        
        //장르 내 play횟수별 정렬
        for(String genre: sortedGenre){
            List<int[]> song = genreInfo.get(genre);

            song.sort((a,b)->{
                if(a[1]==b[1]) return a[0]-b[0];
                else return b[1]-a[1];
            }); 
        }
        
        List<Integer> result = new ArrayList<>();
        
        for(String genre: sortedGenre){
            List<int[]>song = genreInfo.get(genre);
            int cnt=0;
            for(int[] s: song){
                if(cnt<2){
                    result.add(s[0]);
                    cnt++;
                }  
            }
        }
        int[] answer = result.stream().mapToInt(Integer::intValue).toArray();

        return answer;
    }
}

File: 한승우/10월 1주차/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <iostream>

using namespace std;

typedef pair<int, int> ii;
typedef pair<string, int> si;

//장르 -> 재생 횟수
unordered_map<string, int> play_per_genre;
//장르 -> {재생 횟수, 고유 번호}
unordered_map<string, vector<pair<int, int>>> musics;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    
    //초기화
    for(int i = 0; i < genres.size(); i++){
        play_per_genre[genres[i]] += plays[i];
        musics[genres[i]].push_back({plays[i], i});
    }
    
    //정렬
    vector<pair<string, int>> v(play_per_genre.begin(), play_per_genre.end());
    sort(v.begin(), v.end(), [](si a, si b){
        return a.second > b.second;
    });
    
    for(auto& kv: musics){
        vector<pair<int, int>>& v = kv.second;
        sort(v.begin(), v.end(), [](ii a, ii b){
            if(a.first != b.first) return a.first > b.first;
            return a.second < b.second;
        });
    }
    
    //저장
    for(si p: v){
        string genre = p.first;
        
        //음악 두 곡 삽입
        int cnt = 2;
        for(int i = 0; i < musics[genre].size() && cnt > 0; i++){
            answer.push_back(musics[genre][i].second);
            cnt--;
        }
    }
    
    return answer;
}